File: Go/two_sum/1.go
/**
 * Runtime: 7ms (Beats 70.84%)
 * Memory: 4.3MB (Beats 13.87%)
 */

package twosum

func twoSum(nums []int, target int) []int {
	hashMap := make(map[int]int)
	for i, num := range nums {
		needNumber := target - num
		if _, ok := hashMap[needNumber]; ok {
			return []int{hashMap[needNumber], i}
		} else {
			hashMap[num] = i
		}
	}
	return []int{-1}
}

File: TypeScript/two_sum/1.ts
function twoSum(nums: number[], target: number): number[] {
  const map = new Map<number, number>();
  for (let i = 0; i < nums.length; i++) {
    const complement = target - nums[i];
    const value = map.get(complement);
    if (value === undefined) {
      map.set(nums[i], i);
    } else {
      return [value, i];
    }
  }
  return [-1];
}

function twoSum2(nums: number[], target: number): number[] {
  for (let i = 0; i < nums.length; i++) {
    const complement = target - nums[i];
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[j] === complement) {
        return [i, j];
      }
    }
  }
  return [-1];
}

export { twoSum, twoSum2 };
